### ISSUE TITLE
Pods Unable to be created due to being unable to create Endpoints

### ISSUE THREAD

**Issue by solarnz** [cilium/cilium#15428]
## Bug report

**General Information**

- Cilium version (run `cilium version`)

Client: 1.9.5 079bdaf 2021-03-10T13:12:19-08:00 go version go1.15.8 linux/amd64
Daemon: Not responding

- Kernel version (run `uname -a`)

```
Linux ip-172-18-50-73.ap-southeast-2.compute.internal 4.14.219-161.340.amzn2.x86_64 #1 SMP Thu Feb 4 05:54:19 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

- Orchestration system version in use (e.g. `kubectl version`, Mesos, ...)

```
Server Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.9-eks-d1db3c", GitCommit:"d1db3c46e55f95d6a7d3e5578689371318f95ff9", GitTreeState:"clean", BuildDate:"2020-10-20T22:18:07Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
```
Running 1.18 on eks

- Link to relevant artifacts (policies, deployments scripts, ...)

Logs from the cilium agent,

```
level=info msg="Processing API request with rate limiter" name=endpoint-delete parallelRequests=4 subsys=rate uuid=75062f71-8b83-11eb-bee1-0610ca3146ec
level=warning msg="Not processing API request. Wait duration for maximum parallel requests exceeds maximum" error="context canceled" name=endpoint-delete parallelRequests=4 subsys=rate uuid=75062f71-8b83-11eb-bee1-0610ca3146ec
level=info msg="API call has been processed" error="timed out while waiting to be served with 4 parallel requests: context canceled" name=endpoint-delete processingDuration=0s subsys=rate totalDuration=1m30.000121368s uuid=75062f71-8b83-11eb-bee1-0610ca3146ec waitDurationTotal=0s
level=info msg="Processing API request with rate limiter" name=endpoint-delete parallelRequests=4 subsys=rate uuid=ab236f72-8b83-11eb-bee1-0610ca3146ec
level=warning msg="Not processing API request. Wait duration for maximum parallel requests exceeds maximum" error="context canceled" name=endpoint-delete parallelRequests=4 subsys=rate uuid=ab236f72-8b83-11eb-bee1-0610ca3146ec
level=info msg="API call has been processed" error="timed out while waiting to be served with 4 parallel requests: context canceled" name=endpoint-delete processingDuration=0s subsys=rate totalDuration=1m30.00014872s uuid=ab236f72-8b83-11eb-bee1-0610ca3146ec waitDurationTotal=0s
level=info msg="Processing API request with rate limiter" maxWaitDuration=15s name=endpoint-create parallelRequests=4 subsys=rate uuid=e1391011-8b83-11eb-bee1-0610ca3146ec
level=warning msg="Not processing API request. Wait duration for maximum parallel requests exceeds maximum" error="context deadline exceeded" maxWaitDuration=15s name=endpoint-create parallelRequests=4 subsys=rate uuid=e1391011-8b83-11eb-bee1-0610ca3146ec
level=info msg="API call has been processed" error="timed out while waiting to be served with 4 parallel requests: context deadline exceeded" name=endpoint-create processingDuration=0s subsys=rate totalDuration=15.00019634s uuid=e1391011-8b83-11eb-bee1-0610ca3146ec waitDurationTotal=0s
level=info msg="Processing API request with rate limiter" name=endpoint-delete parallelRequests=4 subsys=rate uuid=ea2e132e-8b83-11eb-bee1-0610ca3146ec
level=warning msg="Not processing API request. Wait duration for maximum parallel requests exceeds maximum" error="context canceled" name=endpoint-delete parallelRequests=4 subsys=rate uuid=ea2e132e-8b83-11eb-bee1-0610ca3146ec
level=info msg="API call has been processed" error="timed out while waiting to be served with 4 parallel requests: context canceled" name=endpoint-delete processingDuration=0s subsys=rate totalDuration=1m30.000130175s uuid=ea2e132e-8b83-11eb-bee1-0610ca3146ec waitDurationTotal=0s
level=info msg="Processing API request with rate limiter" name=endpoint-delete parallelRequests=4 subsys=rate uuid=20433800-8b84-11eb-bee1-0610ca3146ec
level=warning msg="Not processing API request. Wait duration for maximum parallel requests exceeds maximum" error="context canceled" name=endpoint-delete parallelRequests=4 subsys=rate uuid=20433800-8b84-11eb-bee1-0610ca3146ec
level=info msg="API call has been processed" error="timed out while waiting to be served with 4 parallel requests: context canceled" name=endpoint-delete processingDuration=0s subsys=rate totalDuration=1m30.000134149s uuid=20433800-8b84-11eb-bee1-0610ca3146ec waitDurationTotal=0s
level=info msg="Processing API request with rate limiter" maxWaitDuration=15s name=endpoint-create parallelRequests=4 subsys=rate uuid=5646d7e5-8b84-11eb-bee1-0610ca3146ec
level=warning msg="Not processing API request. Wait duration for maximum parallel requests exceeds maximum" error="context deadline exceeded" maxWaitDuration=15s name=endpoint-create parallelRequests=4 subsys=rate uuid=5646d7e5-8b84-11eb-bee1-0610ca3146ec
level=info msg="API call has been processed" error="timed out while waiting to be served with 4 parallel requests: context deadline exceeded" name=endpoint-create processingDuration=0s subsys=rate totalDuration=15.000224477s uuid=5646d7e5-8b84-11eb-bee1-0610ca3146ec waitDurationTotal=0s
level=info msg="Processing API request with rate limiter" name=endpoint-delete parallelRequests=4 subsys=rate uuid=5f3bd06e-8b84-11eb-bee1-0610ca3146ec
level=warning msg="Not processing API request. Wait duration for maximum parallel requests exceeds maximum" error="context canceled" name=endpoint-delete parallelRequests=4 subsys=rate uuid=5f3bd06e-8b84-11eb-bee1-0610ca3146ec
level=info msg="API call has been processed" error="timed out while waiting to be served with 4 parallel requests: context canceled" name=endpoint-delete processingDuration=0s subsys=rate totalDuration=1m30.000156491s uuid=5f3bd06e-8b84-11eb-bee1-0610ca3146ec waitDurationTotal=0s
level=info msg="Processing API request with rate limiter" name=endpoint-delete parallelRequests=4 subsys=rate uuid=9566469f-8b84-11eb-bee1-0610ca3146ec
```

Interestingly, `cilium endpoint list` also times out,

```
# cilium endpoint list
Error: cannot get endpoint list: Cilium API client timeout exceeded
```

When creating pods on that particular node, we get

```
Events:
  Type     Reason                  Age                   From     Message
  ----     ------                  ----                  ----     -------
  Warning  FailedCreatePodSandBox  103s (x34 over 109m)  kubelet  (combined from similar events): Failed to create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container "19a749086f762bdf161cba2587e31ad148f9b45b545da5f977cde406c1e46812" network for pod "pelican-5cd8549595-drz7n": networkPlugin cni failed to set up pod "pelican-5cd8549595-drz7n_pelican" network: unable to create endpoint: response status code does not match any response statuses defined for this endpoint in the swagger spec (status 429): {}
```

Spinning up pods on other nodes seems to be fine, and if we restart the cilium agent pod on that particular kubernetes node, it seems to be fine.

- Generate and upload a system zip:
```
curl -sLO https://git.io/cilium-sysdump-latest.zip && python cilium-sysdump-latest.zip
```

[cilium-sysdump-20210323-140231.zip](https://github.com/cilium/cilium/files/6186622/cilium-sysdump-20210323-140231.zip)

**How to reproduce the issue**

I'm not sure on how to reproduce it, but we are consistently getting it every other day in our kubernetes cluster.

**Comment by pchaigno** [comment]
We may have a bug in the rate limiter. /cc @tgraf 

I see four endpoint-delete API requests that are followed by `Delete endpoint request`:
```
2021-03-23T00:37:58.394498080Z level=info msg="Processing API request with rate limiter" name=endpoint-delete parallelRequests=4 subsys=rate uuid=03f563cd-8b70-11eb-bee1-0610ca3146ec
2021-03-23T00:37:58.394519995Z level=info msg="API request released by rate limiter" name=endpoint-delete parallelRequests=4 subsys=rate uuid=03f563cd-8b70-11eb-bee1-0610ca3146ec waitDurationTotal="63.548µs"
2021-03-23T00:37:58.394524220Z level=info msg="Delete endpoint request" id="container-id:3b8947bddc02f426ba6d37d6fc5a28c6b66dfd35d5f20fd30c28560ab4ee13ad" subsys=daemon
```
However, there doesn't seem to be any corresponding `API call has been processed` message. Subsequent endpoint-delete API requests are rejected because there are already 4 requests being processed.

So it's as if the API rate limiter somehow missed the end of the endpoint-delete processing and assumes they are being processed indefinitely.

**Comment by solarnz** [comment]
This might be a completely wrong and misguided comment (please disregard if it is, I don't have any experience with this code base)

What was interesting to me was that `cilium endpoint list` times out,

```
# cilium endpoint list
Error: cannot get endpoint list: Cilium API client timeout exceeded
```

This indicated to me it might be more to do with some underlying lock or resource contention rather than rate limiting?

**Comment by cocatrice** [comment]
Is there any update on this issue @pchaigno @tgraf ?

**Comment by cocatrice** [comment]
I am experiencing the same issue in 1.10.0

- Cilium version: 1.10.0.rc1
- Kernel version: `Linux ip-172-18-46-124 5.4.0-1045-aws #47-Ubuntu SMP Tue Apr 13 07:02:25 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`
- Orchestration system version in use: Using EKS, with `Server Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.9-eks-d1db3c", GitCommit:"d1db3c46e55f95d6a7d3e5578689371318f95ff9", GitTreeState:"clean", BuildDate:"2020-10-20T22:18:07Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}`

Logs in the cilium pod

```
level=warning msg="Not processing API request. Wait duration for maximum parallel requests exceeds maximum" error="context deadline exceeded" maxWaitDuration=15s name=endpoint-create parallelRequests=3 subsys=rate uuid=f81a478b-d3ab-412c-8c7b-d2f13f124c76
level=info msg="API call has been processed" error="timed out while waiting to be served with 3 parallel requests: context deadline exceeded" name=endpoint-create processingDuration=0s subsys=rate totalDuration=15.00054719s uuid=f81a478b-d3ab-412c-8c7b-d2f13f124c76 waitDurationTotal=0s
level=info msg="Processing API request with rate limiter" name=endpoint-delete parallelRequests=4 subsys=rate uuid=3c7978c3-6843-495c-b85b-7a35c2a527cd
level=info msg="API request released by rate limiter" name=endpoint-delete parallelRequests=4 subsys=rate uuid=3c7978c3-6843-495c-b85b-7a35c2a527cd waitDurationTotal="61.222µs"
level=info msg="API call has been processed" error="endpoint not found" name=endpoint-delete processingDuration="2.832µs" subsys=rate totalDuration="75.93µs" uuid=3c7978c3-6843-495c-b85b-7a35c2a527cd waitDurationTotal="61.222µs"
level=info msg="Processing API request with rate limiter" name=endpoint-delete parallelRequests=4 subsys=rate uuid=b1c8e6b6-00cb-4414-ace7-42e403d4e44b
level=info msg="API request released by rate limiter" name=endpoint-delete parallelRequests=4 subsys=rate uuid=b1c8e6b6-00cb-4414-ace7-42e403d4e44b waitDurationTotal="66.027µs"
level=info msg="API call has been processed" error="endpoint not found" name=endpoint-delete processingDuration="2.91µs" subsys=rate totalDuration="81.83µs" uuid=b1c8e6b6-00cb-4414-ace7-42e403d4e44b waitDurationTotal="66.027µs"
level=info msg="Processing API request with rate limiter" maxWaitDuration=15s name=endpoint-create parallelRequests=3 subsys=rate uuid=edfca395-99d8-4949-bfe8-57ecf36c3923
level=warning msg="Not processing API request. Wait duration for maximum parallel requests exceeds maximum" error="context deadline exceeded" maxWaitDuration=15s name=endpoint-create parallelRequests=3 subsys=rate uuid=edfca395-99d8-4949-bfe8-57ecf36c3923
level=info msg="API call has been processed" error="timed out while waiting to be served with 3 parallel requests: context deadline exceeded" name=endpoint-create processingDuration=0s subsys=rate totalDuration=15.001201443s uuid=edfca395-99d8-4949-bfe8-57ecf36c3923 waitDurationTotal=0s
level=info msg="Processing API request with rate limiter" name=endpoint-delete parallelRequests=4 subsys=rate uuid=d207fea3-e98b-4b47-88d1-d872afa251e2
level=info msg="API request released by rate limiter" name=endpoint-delete parallelRequests=4 subsys=rate uuid=d207fea3-e98b-4b47-88d1-d872afa251e2 waitDurationTotal="59.936µs"
level=info msg="API call has been processed" error="endpoint not found" name=endpoint-delete processingDuration="2.944µs" subsys=rate totalDuration="74.803µs" uuid=d207fea3-e98b-4b47-88d1-d872afa251e2 waitDurationTotal="59.936µs"
level=info msg="Processing API request with rate limiter" name=endpoint-delete parallelRequests=4 subsys=rate uuid=58932521-9a9b-4131-a2f9-c17040df81d5
level=info msg="API request released by rate limiter" name=endpoint-delete parallelRequests=4 subsys=rate uuid=58932521-9a9b-4131-a2f9-c17040df81d5 waitDurationTotal="61.334µs"
level=info msg="API call has been processed" error="endpoint not found" name=endpoint-delete processingDuration="3.449µs" subsys=rate totalDuration="76.034µs" uuid=58932521-9a9b-4131-a2f9-c17040df81d5 waitDurationTotal="61.334µs"
level=info msg="Processing API request with rate limiter" maxWaitDuration=15s name=endpoint-create parallelRequests=3 subsys=rate uuid=94c784f6-4e20-4de7-8727-6c250bb7cbc3
```

No pod is able to be able to scheduled on that node. 

```
➜  ~ k exec -it cilium-rq95g  -n kube-system cilium endpoint list
kubectl exec [POD] [COMMAND] is DEPRECATED and will be removed in a future version. Use kubectl exec [POD] -- [COMMAND] instead.
Defaulted container "cilium-agent" out of: cilium-agent, wait-for-node-init (init), clean-cilium-state (init)
Error: cannot get endpoint list: Cilium API client timeout exceeded

command terminated with exit code 1
➜  ~ 
```

Delete the cilium pod on the node seems to be working again.

**Comment by chingyi-lin** [comment]
I'm getting it on EKS 1.19 with `aws-cni` chain mode as well, and yes deleting the cilium pod on the node makes it work again.
sysdump: [cilium-sysdump-20210625-114341.zip](https://github.com/cilium/cilium/files/6718233/cilium-sysdump-20210625-114341.zip)

**Comment by christarazi** [comment]
Looks like a deadlock from the sysdump:

```
goroutine 20725 [semacquire, 108 minutes]:
sync.runtime_SemacquireMutex(0xc00037fb3c, 0x406400, 0x0)
	/usr/local/go/src/runtime/sema.go:71 +0x47
sync.(*RWMutex).RLock(...)
	/usr/local/go/src/sync/rwmutex.go:50
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).unconditionalRLock(...)
	/go/src/github.com/cilium/cilium/pkg/endpoint/lock.go:60
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).HasLabels(0xc00037fb00, 0x0, 0x0)
	/go/src/github.com/cilium/cilium/pkg/endpoint/endpoint.go:1018 +0xb3
github.com/cilium/cilium/daemon/cmd.(*Daemon).getEndpointList.func1(0xc00225e068, 0xc000ec1070, 0xc004503800, 0xc0045037a0)
	/go/src/github.com/cilium/cilium/daemon/cmd/endpoint.go:107 +0x75
created by github.com/cilium/cilium/daemon/cmd.(*Daemon).getEndpointList
	/go/src/github.com/cilium/cilium/daemon/cmd/endpoint.go:105 +0x1c5

goroutine 26571 [semacquire, 96 minutes]:
sync.runtime_SemacquireMutex(0xc00037f234, 0xc004e9c400, 0x1)
	/usr/local/go/src/runtime/sema.go:71 +0x47
sync.(*Mutex).lockSlow(0xc00037f230)
	/usr/local/go/src/sync/mutex.go:138 +0x105
sync.(*Mutex).Lock(...)
	/usr/local/go/src/sync/mutex.go:81
sync.(*RWMutex).Lock(0xc00037f230)
	/usr/local/go/src/sync/rwmutex.go:98 +0x90
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).lockAlive(0xc00037f200, 0x289bf60, 0xc004e9c5e8)
	/go/src/github.com/cilium/cilium/pkg/endpoint/lock.go:21 +0x3a
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).SetRegenerateStateIfAlive(0xc00037f200, 0xc00162e6c0, 0x0, 0x0, 0xc004e9c640)
	/go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:517 +0x2f
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).RegenerateIfAlive(0xc00037f200, 0xc00162e6c0, 0xc004e9c7b8)
	/go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:556 +0x50
github.com/cilium/cilium/pkg/endpointmanager.(*EndpointManager).RegenerateAllEndpoints.func1(0xc00162e6c0, 0xc00221f0c0, 0xc00037f200)
	/go/src/github.com/cilium/cilium/pkg/endpointmanager/manager.go:435 +0x35
created by github.com/cilium/cilium/pkg/endpointmanager.(*EndpointManager).RegenerateAllEndpoints
	/go/src/github.com/cilium/cilium/pkg/endpointmanager/manager.go:434 +0x225
```

**Comment by othmane399** [comment]
Hello @ldelossa, any news on this issue ? we are facing the same behavior on some of our clusters using cilium v1.10.7

**Comment by aanm** [comment]
@solarnz @othmane399 can you check if this is fixed with v1.10.10? Thank you

**Comment by ksramchandani** [comment]
@aanm Can you please provide the details of the 1.9 version where the fix was backported?
